src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -964,50 +956,33 @@ internal static void CreateNameToMembersMap
         {
             var dictionary = new Dictionary<string, ImmutableArray<TNamedTypeSymbol>>(comparer);
 
-            foreach (var kvp in map)
-            {
-                ImmutableArray<TNamespaceOrTypeSymbol> members = kvp.Value;
+            foreach (var (name, members) in map)
+                dictionary.Add(name, getOrCreateNamedTypes(members));
 
-                bool hasType = false;
-                bool hasNamespace = false;
+            return dictionary;
 
-                foreach (var symbol in members)
+            ImmutableArray<TNamedTypeSymbol> getOrCreateNamedTypes(ImmutableArray<TNamespaceOrTypeSymbol> members)
+            {
+                // See if creator 'map' put a downcasted ImmutableArray<TNamedTypeSymbol> in it.  If so, we can just directly
+                // downcast to that and trivially reuse it.  If not, that means the array must have contained at least one
+                // TNamespaceSymbol and we'll need to filter that out.
+                var membersAsNamedTypes = members.As<TNamedTypeSymbol>();
```

Comment:
we know that if we only had named types then CreateNameToMembersMap already put in a downcasted IA, so we can just directly try to move to that.  no need for walking the arrays *again*.